gemcalibration/src/common/Calibration.cc

```diff
@@ -0,0 +1,398 @@
+/**
+ * class: Calibration
+ * description: Calibration application for GEM system
+ *              structure borrowed from gemSupervisor
+ * author: 
+ * date:
+ */
+
+#include "gem/calib/Calibration.h"
+
+#include <cstdlib>
+#include <iomanip>
+
+#include <map>
+#include <set>
+#include <vector>
+#include <algorithm>
+
+#include <boost/algorithm/string.hpp>
+
+#include "gem/calib/CalibrationWeb.h"
+
+//#include "gem/utils/soap/GEMSOAPToolBox.h"
+//#include "gem/utils/exception/Exception.h"
+
+typedef gem::base::utils::GEMInfoSpaceToolBox::UpdateType GEMUpdateType;
+
+XDAQ_INSTANTIATOR_IMPL(gem::calib::Calibration);
+
+gem::calib::Calibration::Calibration(xdaq::ApplicationStub* stub) :
+    gem::base::GEMApplication(stub),
+    m_nShelves(0)
+{
+    CMSGEMOS_DEBUG("gem::calib::Calibration : Creating the CalibrationWeb interface");
+    CMSGEMOS_DEBUG("gem::calib::Calibration : Retrieving configuration");
+    p_appInfoSpace->fireItemAvailable("nShelves",     &m_nShelves);
+    p_appInfoSpace->addItemRetrieveListener("nShelves", this);
+    p_appInfoSpace->addItemChangedListener("nShelves", this);
+    CMSGEMOS_DEBUG("gem::calib::Calibration : configuration retrieved");
+    CMSGEMOS_INFO("gem::calib::Calibration:init() number of shelves: " << m_nShelves.value_);
+    p_gemWebInterface = new gem::calib::CalibrationWeb(this);
+    m_calType = NDEF;
+  
+    xgi::bind(this, &Calibration::applyAction, "applyAction");
+    xgi::bind(this, &Calibration::setCalType, "setCalType");
+
+   
+}
+
+gem::calib::Calibration::~Calibration()
+{
+    CMSGEMOS_DEBUG("gem::calib::Calibration : Destructor called");
+    // make sure to empty the v_supervisedApps  vector and free the pointers
+}
+
+void gem::calib::Calibration::actionPerformed(xdata::Event& event)
+{
+    if (event.type() == "setDefaultValues" || event.type() == "urn:xdaq-event:setDefaultValues") {
+        CMSGEMOS_DEBUG("gem::calib::Calibration::actionPerformed() setDefaultValues" <<
+                       "Default configuration values have been loaded from xml profile");
+        init();
+    }
+
+    // item is changed, update it
+    if (event.type() == "ItemChangedEvent" || event.type() == "urn:xdata-event:ItemChangedEvent") {
+        CMSGEMOS_DEBUG("gem::calib::Calibration::actionPerformed() ItemChangedEvent");
+    }
+    // update calibration variables
+    gem::base::GEMApplication::actionPerformed(event);
+}
+
+void gem::calib::Calibration::init()
+{
+    CMSGEMOS_INFO("gem::calib::Calibration:init() number of shelves: " << m_nShelves.value_);
+ 
+}
+
+void gem::calib::Calibration::applyAction(xgi::Input* in, xgi::Output* out)
+{
+    bool t_errorsOccured = false;
+    cgicc::Cgicc cgi(in);
+      
+    std::map<std::string, std::string >::iterator it = (m_scanParams.find(m_calType)->second).begin();
```

Comment:
I'm not sure I like the way `m_calType` is used, i.e. it adds a state to the xDAQ web server application. What happens if some one opens twice the calibration suite webpage and selects a different calibration type in each of them?

Reply:
The initial design supposed that the whole management of the calibration run will happen from the calibration web page. In this case, the internal state would be reflected on the webpage instantly with ajax and parts of interface will be frozen accordingly to the state. I.e. if simultaneous edits happen, the state would be unique and correspond to the latest edit. But if the "start" button is pressed, the parameter interface is frozen until the scan is finished. However this behavior is not possible with the current layout where `GEMSupervisor` drives the data taking and calibration page only serves the parameters. Also interface freezing has not been implemented. Good candidate for redesign.